go/libraries/doltcore/sqle/stats/auto_refresh.go

```diff
@@ -0,0 +1,303 @@
+							curStats.latestTableHashes[table] = tableHash
+						}
+					}
+					// get new buckets for index chunks to update
+					newTableStats, err := updateStats(sqlCtx, sqlTable, dTab, indexes, idxMetas)
+					if err != nil {
+						sqlCtx.GetLogger().Debugf("statistics refresh error: %s", err.Error())
+						continue
+					}
+
+					// merge new chunks with preexisting chunks
+					for _, idxMeta := range idxMetas {
+						stat := newTableStats[idxMeta.qual]
+						if stat != nil {
+							newStats[idxMeta.qual] = mergeStatUpdates(stat, idxMeta)
+						}
+					}
+				}
+
+				for _, s := range curStats.stats {
+					// table or index delete leaves hole in stats
+					// this is separate from threshold check
+					if !tableExistsAndSkipped[s.Qual.Table()] && !qualExists[s.Qual] {
+						// only delete stats we've verified are deleted
+						deletedStats = append(deletedStats, s.Qual)
+					}
+				}
+
+				prevMap := curStats.currentMap
+				if prevMap.KeyDesc().Count() == 0 {
+					kd, vd := schema.StatsTableDoltSchema.GetMapDescriptors()
+					prevMap, err = prolly.NewMapFromTuples(ctx, ddb.NodeStore(), kd, vd)
+					if err != nil {
+						sqlCtx.GetLogger().Debugf("statistics refresh error: %s", err.Error())
+						continue
+					}
+				}
+
+				if len(deletedStats) > 0 {
+					sqlCtx.GetLogger().Debugf("statistics refresh: deleting stats %#v", deletedStats)
+				}
+				delMap, err := deleteStats(sqlCtx, prevMap, deletedStats...)
+				if err != nil {
+					sqlCtx.GetLogger().Debugf("statistics refresh error: %s", err.Error())
+					continue
+				}
+
+				newMap, err := flushStats(sqlCtx, delMap, newStats)
+				if err != nil {
+					sqlCtx.GetLogger().Debugf("statistics refresh error: %s", err.Error())
+					continue
+				}
+
+				p.mu.Lock()
+				p.dbStats[dbName].currentMap = newMap
+				err = ddb.SetStatisics(ctx, newMap.HashOf())
+				for q, s := range newStats {
+					p.dbStats[dbName].stats[q] = s
+				}
+				p.mu.Unlock()
+				if err != nil {
+					sqlCtx.GetLogger().Debugf("statistics refresh error: %s", err.Error())
+					continue
+				}
+			}
+		}
+	})
+}
+
+func newIdxMeta(ctx *sql.Context, curStats *DoltStats, doltTable *doltdb.Table, sqlIndex sql.Index, cols []string) (indexMeta, error) {
+	var idx durable.Index
+	var err error
+	if strings.EqualFold(sqlIndex.ID(), "PRIMARY") {
+		idx, err = doltTable.GetRowData(ctx)
+	} else {
+		idx, err = doltTable.GetIndexRowData(ctx, sqlIndex.ID())
+	}
+	if err != nil {
+		return indexMeta{}, err
+	}
+
+	prollyMap := durable.ProllyMapFromIndex(idx)
+
+	// get histogram level hashes
+	levelNodes, err := tree.GetHistogramLevel(ctx, prollyMap.Tuples(), bucketLowCnt)
+	if err != nil {
+		return indexMeta{}, err
+	}
+
+	var addrs []hash.Hash
+	var missingAddrs float64
+	var missingChunks []tree.Node
+	var preservedStats []DoltBucket
+	var missingOffsets [][]uint64
+	var offset uint64
+	for _, n := range levelNodes {
+		// check if hash is in current list
```

Comment:
Can you add a comment explaining what's going on in this block?

Reply:
Added a comment. Basically partition the target level histogram hashes for the current tree into 1) preserved or 2) missing, and track scanning metadata for missing chunks

Reply:
Okay I follow. Is there a particular reason we do it this way by storing all the hashes at some level of the tree, instead of just storing the root hash?

Reply:
Maybe I'm misunderstanding the question, but each histogram bucket is the contexts of a chunk at the target level, one-to-one equivalence. We do track the table hash to short-circuit the target level comparison if nothing has changed, but otherwise we use structural sharing to 1) get an estimate of how much has changed, and 2) avoid unnecessarily recomputing buckets that haven't changed.

Reply:
Ah, okay. I missed that the chunks being tracked here mapped 1-1 with histogram buckets.

This makes sense.

Reply:
WIil update docstrings to clarify